### DISCUSSION TITLE
How to blacklist everything in a directory except one (or more) sub-directories

### DISCUSSION THREAD

**Discussion by YorkZ** [netblue30/firejail#5513] [Q&A] [answered]
Let's say I have the following directories:

- ${HOME}/devices
- ${HOME}/devices/device1/audio 
- ${HOME}/devices/device1/otherstuff
- ${HOME}/devices/device2
- ${HOME}/devices/device3

I started `vlc`, it didn't have access to `${HOME}/devices` at all. So I added the following:

```
whitelist ${HOME}/devices 
```

to the file `${HOME}/.config/firejail/vlc.local`, and then `vlc` was able to access the entire directory `${HOME}/devices`. However, I want `vlc` to be able to play audio files in `${HOME}/devices/device1/audio` only, but blacklist everything else in `${HOME}/devices`. I've tried a few combinations with `backlist`, `noblacklist` etc, but none of them achieved my goal. Please help, thank you very much.

**Comment by kmk3** [comment]
@YorkZ [on Dec 8](https://github.com/netblue30/firejail/discussions/5513#discussion-4639912):

> Let's say I have the following directories:
> 
> * ${HOME}/devices
> * ${HOME}/devices/device1/audio
> * ${HOME}/devices/device1/otherstuff
> * ${HOME}/devices/device2
> * ${HOME}/devices/device3
> 
> I started `vlc`, it didn't have access to `${HOME}/devices` at all. So I
> added the following:
> 
> ```
> whitelist ${HOME}/devices 
> ```
> 
> to the file `${HOME}/.config/firejail/vlc.local`, and then `vlc` was able to
> access the entire directory `${HOME}/devices`. However, I want `vlc` to be
> able to play audio files in `${HOME}/devices/device1/audio` only, but
> blacklist everything else in `${HOME}/devices`. I've tried a few combinations
> with `backlist`, `noblacklist` etc, but none of them achieved my goal. Please
> help, thank you very much.

Are there any other commands being used in included profiles (such as
globals.local) that affects these directories?

If not, it should be as simple as:

```firejail
whitelist ${HOME}/devices/device1/audio
```

Otherwise, maybe something like this:

```firejail
noblacklist ${HOME}/devices/device1
noblacklist ${HOME}/devices/device1/audio

blacklist ${HOME}/devices/*
blacklist ${HOME}/devices/device1/*

mkdir ${HOME}/devices/device1/audio
whitelist ${HOME}/devices/device1/audio
```

Note: The `mkdir` is optional.

If it works, then the `blacklist` commands could be put in globals.local (to
restrict the directories in question in all profiles) and the rest into
vlc.local, for example.

**Comment by YorkZ** [comment]
Thank you very much @kmk3.

```
whitelist ${HOME}/devices/device1/audio
```

didn't work. But:

```
noblacklist ${HOME}/devices/device1
noblacklist ${HOME}/devices/device1/audio

blacklist ${HOME}/devices/*
blacklist ${HOME}/devices/device1/*

mkdir ${HOME}/devices/device1/audio
whitelist ${HOME}/devices/device1/audio
```

worked a lot "better" in that now the directory `${HOME}/devices/device1/audio` is visible by `vlc`. However, `vlc` wasn't able to see any files in this directory even though there are quite a few `.m4a` files there.

**Reply by rusty-snake** [reply]
> whitelist ${HOME}/devices/device1/audio
> didn't work.

Did you removed the whitelist from OP?

**Reply by YorkZ** [reply]
Yes, I did.

**Reply by rusty-snake** [reply]
> However, vlc wasn't able to see any files in this directory even though there are quite a few .m4a files there.

Is the directory empty or inaccessible by vlc?

**Reply by rusty-snake** [reply]
Is there anything "special" about `devices`?

**Reply by kmk3** [reply]
@YorkZ [on Dec 8](https://github.com/netblue30/firejail/discussions/5513#discussioncomment-4346179):

> Thank you very much @kmk3.

No problem.

> ```
> whitelist ${HOME}/devices/device1/audio
> ```
> 
> didn't work.

It works for me with only that command:

<details>
<summary>Attempt to reproduce</summary>
<p>

```console
# Init
mkdir -p "${HOME}/devices/device1/audio"
mkdir -p "${HOME}/devices/device1/otherstuff"
mkdir -p "${HOME}/devices/device2"
mkdir -p "${HOME}/devices/device3"

touch "${HOME}/devices/device1/audio/file"
touch "${HOME}/devices/device1/otherstuff/file"
touch "${HOME}/devices/device2/file"
touch "${HOME}/devices/device3/file"

# Whitelisting ~/devices
$ firejail --quiet --ignore=private-bin \
  --ignore='include vlc.local' --ignore='include globals.local' \
  --whitelist="${HOME}/devices" \
  --profile=vlc find devices | LC_ALL=C sort
devices
devices/device1
devices/device1/audio
devices/device1/audio/file
devices/device1/otherstuff
devices/device1/otherstuff/file
devices/device2
devices/device2/file
devices/device3
devices/device3/file

# Whitelisting ~/devices/device1/audio
$ firejail --quiet --ignore=private-bin \
  --ignore='include vlc.local' --ignore='include globals.local' \
  --whitelist="${HOME}/devices/device1/audio" \
  --profile=vlc find devices -type f | LC_ALL=C sort
devices
devices/device1
devices/device1/audio
devices/device1/audio/file

# Same, but showing files only
$ firejail --quiet --ignore=private-bin \
  --ignore='include vlc.local' --ignore='include globals.local' \
  --whitelist="${HOME}/devices/device1/audio" \
  --profile=vlc find devices -type f | LC_ALL=C sort
devices/device1/audio/file
```

</p>
</details>

This is with firejail-git (d6b0d40a1) on Artix Linux.

What OS and firejail version are you using?

Can you reproduce with these exact same commands?

Does ~/.config/firejail/vlc.profile exist?  (It shouldn't in most cases)

Did you change /etc/firejail/vlc.profile at all?

> But:
> 
> ```
> noblacklist ${HOME}/devices/device1
> noblacklist ${HOME}/devices/device1/audio
> 
> blacklist ${HOME}/devices/*
> blacklist ${HOME}/devices/device1/*
> 
> mkdir ${HOME}/devices/device1/audio
> whitelist ${HOME}/devices/device1/audio
> ```
> 
> worked a lot "better" in that now the directory
> `${HOME}/devices/device1/audio` is visible by `vlc`. However, `vlc` wasn't
> able to see any files in this directory even though there are quite a few
> `.m4a` files there.

I wonder if .m4a files even show up in the file picker by default (as this file
extension is not as common as .mp4) or if the behavior changes if some
file-picker-related files are unavailable.

Do the .m4a files appear if you select "all files" in the file picker?

**Comment by YorkZ** [comment, marked as answer]
Thank you so much @kmk3 and @rusty-snake. I have to apologize that this turned out to be my stupid mistake. The problem was that the directory name on the file system was `device1`, but what I had in the file `vlc.local` was:

```
mkdir ${HOME}/devices/device-1/audio
whitelist ${HOME}/devices/device-1/audio
```
I apologize again for having wasted your time!

On the other hand, I actually learned something from you because your testing script is really interesting, and it inspired me that I could use this approach in the future for troubleshooting and testing. Thanks again for this.

By the way, I use Arch Linux with firejail-git e52f5e002.

**Reply by kmk3** [reply]
It's alright; glad the script helped :)
